text/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,520 @@
+* **Enable error handling with `longjmp`:**
+  As mentioned above, some existing Rust libraries rely on the ability to
+  `longjmp` across Rust frames to interoperate with Ruby, Lua, and other C
+  APIs. The behavior of `longjmp` traversing Rust frames is not specified or
+  guaranteed to be safe; in the current implementation of `rustc`,
+  however, it [is safe][longjmp-pr]. On Windows, `longjmp` is implemented as a
+  form of unwinding called ["forced unwinding"][forced-unwinding], so any
+  specification of the behavior of forced unwinding across FFI boundaries
+  should be forward-compatible with a [future RFC][unresolved-questions] that
+  will provide a well-defined way to interoperate with longjmp-based APIs.
+* **Do not change the ABI of functions in the `libc` crate:** Some `libc`
+  functions may invoke `pthread_exit`, which uses [a form of
+  unwinding][forced-unwinding] in the GNU libc implementation. Such functions
+  must be safe to use with the existing `"C"` ABI, because changing the types
+  of these functions would be a breaking change. 
+
+[inside-rust-requirements]: https://blog.rust-lang.org/inside-rust/2020/02/27/ffi-unwind-design-meeting.html#requirements-for-any-cross-language-unwinding-specification
+[longjmp-pr]: https://github.com/rust-lang/rust/pull/48572
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring an external function that may unwind, such as an entrypoint to a
+C++ library, use `extern "C-unwind"` instead of `extern "C"`:
+
+```
+extern "C-unwind" {
+  fn may_throw();
+}
+```
+
+Rust functions that call a possibly-unwinding external function should either
+use the default Rust ABI (which can be made explicit with `extern "Rust"`) or
+the `"C-unwind"` ABI:
+
+```
+extern "C-unwind" fn can_unwind() {
+  may_throw();
+}
+```
+
+Using the `"C-unwind"` ABI to "sandwich" Rust frames between frames from
+another language (such as C++) allows an exception initiated in a callee frame
+in the other language to traverse the intermediate Rust frames before being
+caught in the caller frames. I.e., a C++ exception may be thrown,
+cross into Rust via an `extern "C-unwind"` function declaration, safely unwind
+the Rust frames, and cross back into C++ (where it may be caught) via a Rust
+`"C-unwind"` function definition.
+
+Conversely, languages that support the native unwinding mechanism, such as C++,
+may be "sandwiched" between Rust frames, so that Rust `panic`s may safely
+unwind the C++ frames, if the Rust code declares both the C++ entrypoint and
+the Rust entrypoint using `"C-unwind"`.
+
+## Other `unwind` ABI strings
+
+Because the `C` ABI is not appropriate for all use cases, we also introduce
+these `unwind` ABI strings, which will only differ from their non-`unwind`
+variants by permitting unwinding, with the same semantics as `"C-unwind"`:
+
+* `"system unwind"` - available on all platforms
+* `"stdcall unwind"` and `"thiscall unwind"` - available only on platforms
```

Comment:
Out of curiosity, if `"C unwind"` became `"C-unwind"`, shouldn't `"system unwind"`, `"stdcall unwind"`, `"thiscall unwind"` become `"system-unwind"`, `"stdcall-unwind"`, `"thiscall-unwind"` for consistency?

Reply:
Oops, you are absolutely correct! Fixed.